src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -743,110 +748,205 @@ private enum EnumeratorResult
+                WellKnownType wellKnownSpan;
+                bool usedAsValue = false;
+
+                if (CheckValueKind(collectionExpr.Syntax, collectionExpr, BindValueKind.RefersToLocation | BindValueKind.Assignable, checkingReceiver: false, BindingDiagnosticBag.Discarded))
                 {
-                    // Anything else with a null type is a method group or anonymous function
-                    diagnostics.Add(ErrorCode.ERR_AnonMethGrpInForEach, _syntax.Expression.Location, collectionExpr.Display);
+                    wellKnownSpan = WellKnownType.System_Span_T;
                 }
+                else
+                {
+                    wellKnownSpan = WellKnownType.System_ReadOnlySpan_T;
 
-                // CONSIDER: dev10 also reports ERR_ForEachMissingMember (i.e. failed pattern match).
-                return EnumeratorResult.FailedAndReported;
-            }
+                    if (!CheckValueKind(collectionExpr.Syntax, collectionExpr, BindValueKind.RefersToLocation, checkingReceiver: false, BindingDiagnosticBag.Discarded))
+                    {
+                        usedAsValue = true;
+                    }
+                }
 
-            if (collectionExpr.ResultKind == LookupResultKind.NotAValue)
-            {
-                // Short-circuiting to prevent strange behavior in the case where the collection
-                // expression is a type expression and the type is enumerable.
-                Debug.Assert(collectionExpr.HasAnyErrors); // should already have been reported
-                return EnumeratorResult.FailedAndReported;
-            }
+                NamedTypeSymbol spanType = GetWellKnownType(wellKnownSpan, diagnostics, collectionExpr.Syntax);
 
-            if (collectionExprType.Kind == SymbolKind.DynamicType && IsAsync)
-            {
-                diagnostics.Add(ErrorCode.ERR_BadDynamicAwaitForEach, _syntax.Expression.Location);
-                return EnumeratorResult.FailedAndReported;
+                if (spanType.IsErrorType())
+                {
+                    builder = new ForEachEnumeratorInfo.Builder();
+                    return EnumeratorResult.FailedAndReported;
+                }
+
+                spanType = spanType.Construct(ImmutableArray.Create(elementField.TypeWithAnnotations));
+                spanType.CheckConstraints(new ConstraintsHelper.CheckConstraintsArgs(this.Compilation, this.Conversions, collectionExpr.Syntax.GetLocation(), diagnostics));
+
+                var enumeratorInfoDiagnostics = BindingDiagnosticBag.GetInstance(diagnostics);
+                BoundExpression span = new BoundValuePlaceholder(collectionExpr.Syntax, spanType).MakeCompilerGenerated();
+#if DEBUG
+                var originalSpan = span;
+#endif
+
+                result = getEnumeratorInfo(ref span, isAsync: false, enumeratorInfoDiagnostics, out builder);
+
+#if DEBUG
+                Debug.Assert(span == originalSpan);
```

Comment:
What if `Span<T>` is defined as a nullable `class`, could this assert fail then?

Reply:
> What if ```Span<T>``` is defined as a nullable class, could this assert fail then?

No, ```IsNullableType``` returns true inly for ```System.Nullable<T>```. The ```span``` can never have type like that.